### DISCUSSION TITLE
Revalidate when children changes

### DISCUSSION THREAD

**Discussion by simplecommerce** [vazco/uniforms#1139] [Q&A] [answered]
Hi, 

I was wondering if you guys had any better approach/suggestion.

I have a form with a Select component.
When I select an option, I render other fields that are required based on the selected option.

My issue, is when the new fields render and appear in the form and `submitted` is `true`, it will not re-execute the validation and won't include the new fields.

The only way I found to make this work was to add a `useEffect` in my children components and call `validate()` when they mount if `submitted` is `true`.

If you have any ideas or better approach let me know.

**Comment by radekmie** [comment, marked as answer]
Hi @simplecommerce. Technically, it shouldn't matter if a field is visible or not - the user completed it by successfully validating the data. In other words, the visibility of a field should not affect its validation.

What we do in our [Form Builder](https://www.vazco.eu/form-builder) is a schema that is aware of visibility, i.e., both the validation and rendering use the same logic. This makes this problem non-existent as if a field should not be visible, it is not validated and it gets validated if it is. The last part is that the conditions may include only one source of variables - the state of the form (i.e., values of other inputs).

What are the conditions you show/hide your fields on? If they rely only on the model, it should be pretty straightforward.

**Reply by simplecommerce** [reply]
@radekmie I will try to explain better what I was doing.

I am using GraphQL as the bridge for my form.
I had a schema with a field called `typeData: JSON` which is a custom scalar.

In my custom bridge, I auto validate required fields. I have a custom method in my bridge that I call in `onValidate`.
Since `typeData` is a custom scalar and we have no idea of the underlying fields, I had to modify the bridge, to return Object when the name is `JSON` (matching the scalar).

I modified `getProps` to set an internal variable called `fieldProps` if a field `required` prop is set to `true`.

In my form I have a Select field which is called `type`, based on the type, I render different fields.
When I render the fields, if they are required, I put `required` prop on them.
The name of the fields, are prefixed with `typeData`.

So far example, if `type` is `PAYMENT`.
I render a field called `typeData.date` and add a `required` prop to it.

This allows me to mark this field as required and when I hit submit, the validator will pick it up and give an error if it is missing.

The problem with this approach, is that if I do not select a type and hit submit. The validator kicks in and tells me to select a type.
When I select the type `PAYMENT` and the new field `date` is rendered, `onValidate` isn't re-executed, so it doesn't see the new fields and doesn't indicate it as being required.

Sorry if I am being unclear, but this was where I was stuck.
Currently, I changed my whole approach and went with static schemas for every type.

But if I could've figured out how to fix the issue above, I could've kept it dynamic.

Thanks for the time!

**Reply by simplecommerce** [reply]
I will make a repo to demonstrate my issue so it will be easier to see what I am trying to achieve.

**Reply by radekmie** [reply]
> When I select the type PAYMENT and the new field date is rendered, onValidate isn't re-executed [...]

But why? If it would, the problem is solved. You could have these "differentiator fields" (I like to call them ["kinds"](https://en.wikipedia.org/wiki/Kind_(type_theory))) in the schema, and remove them when the form is submitted. You could even fit them into a `$` or `__switches__` field to make them "more separate" (for people, of course).

Other than that, I'd like to see the repo to play with it myself.

**Reply by simplecommerce** [reply]
> > When I select the type PAYMENT and the new field date is rendered, onValidate isn't re-executed [...]
> 
> But why? If it would, the problem is solved. You could have these "differentiator fields" (I like to call them ["kinds"](https://en.wikipedia.org/wiki/Kind_(type_theory))) in the schema, and remove them when the form is submitted. You could even fit them into a `$` or `__switches__` field to make them "more separate" (for people, of course).
> 
> Other than that, I'd like to see the repo to play with it myself.

I wasn't able to create a working repo to reproduce the issue as of now, because I think my logic for dynamic field (JSON) is flawed.
So for now I am leaving this aside.

I did however encounter an issue with the same principle of revalidating when a children changes.
In order to dynamically require a field that isn't required in my schema, I am passing the `required` prop to the child in question.
In my custom bridge, in the `getProps` method, I am storing the field's required prop in a local variable in the class.

In my custom method which validates all required field based on schema, had it also look at the local variable in order to validate it, if it isn't required in the schema but might be required because of the prop.

My validaton method is called using `onValidate`

The goal of this scenario was to be able to force a field to be required based on a selection or input from another field.

My problem seems to be when the form is in a `validating` state.
It calls the `onValidate` before the children are re-rendered. In my case since the `required` prop changed in a child, it doesn't see it and only sees the previous value.

In order to fix this, I was forced to use `useEffect` in my child to listen for a change in the `required` prop and call `validate()`.

Here is the repo I was working on but I tweaked for this specific scenario.

https://codesandbox.io/s/graphqlbridge-with-dynamic-field-json-scalar-is9nnr?file=/src/App.js

I know this isn't really an issue with your library, but I was wondering if you see any other way around this.

Thanks!

**Reply by radekmie** [reply]
Honestly, I've never thought of a case, where passing `required` as a prop to a field would mean this field is actually required in the validation. It's because the bridges (schemas) are stateless by default, and should not care about the UI at all - the data is the only thing that matters.

We did have a few times a similar case, and we solved it by making a function that created a new bridge based on some conditions. Then we memoized all of them and used them as usual.

To be honest, I don't have a solution now, but using this `useEffect` is not bad, I think. You could even have one `useEffect`, next to the place where the form is rendered, that could have both the previous and current model and do that in there. Alternatively, you could schedule a validation in `onChangeModel`.

**Reply by simplecommerce** [reply]
> Honestly, I've never thought of a case, where passing `required` as a prop to a field would mean this field is actually required in the validation. It's because the bridges (schemas) are stateless by default, and should not care about the UI at all - the data is the only thing that matters.
> 
> We did have a few times a similar case, and we solved it by making a function that created a new bridge based on some conditions. Then we memoized all of them and used them as usual.
> 
> 


To be honest, I don't have a solution now, but using this `useEffect` is not bad, I think. You could even have one `useEffect`, next to the place where the form is rendered, that could have both the previous and current model and do that in there. Alternatively, you could schedule a validation in `onChangeModel`.

yeah I know my use case is very specific and I wasn't expecting any other solutions at this moment, but I appreciate your reply.
I will try and check if `onChangeModel` can work for my case also. Your function to create a new bridge, did it create a new one when a model value changed? Or did you generate every possible bridges based on possible values?

Thanks again!

**Reply by radekmie** [reply]
> Your function to create a new bridge, did it create a new one when a model value changed? Or did you generate every possible bridges based on possible values?

It took a few flags from the model, like "x is bigger than 100" or "both z and y are present", and was run every time, but in a memoized function.
